### PR TITLE
iterate over nonzeros

### PR DESCRIPTION
**Pull request by GiggleLiu**
* remove the type parameter `N` of IMatrix.
* add a new API `IterNz` to iterate over nonzero entries of a matrix.

### CODE REVIEW

File: src/linalg.jl [hunk 1]
```diff
@@ -1,25 +1,21 @@
...
 #det(M::PermMatrix) = parity(M.perm)*prod(M.vals)
-function inv(M::PermMatrix)
+function Base.inv(M::PermMatrix)
     new_perm = fast_invperm(M.perm)
     return PermMatrix(new_perm, 1.0 ./ M.vals[new_perm])
 end
 
 ####### multiply ###########
-*(A::IMatrix{N}, B::AbstractVector) where {N} =
+Base.:*(A::IMatrix, B::AbstractVector) =
```
Comment (Roger-luo) [inline src/linalg.jl:18]: can we overload `mul!` this time?

File: src/linalg.jl [hunk 2]
```diff
@@ -129,16 +123,16 @@ end
...
 for MAT in
     [:AbstractArray, :AbstractVector, :Matrix, :SparseMatrixCSC, :PermMatrix, :IMatrix]
-    @eval *(A::Adjoint{<:Any,<:$MAT}, D::IMatrix) = Adjoint(D * parent(A))
-    @eval *(A::Transpose{<:Any,<:$MAT}, D::IMatrix) = Transpose(D * parent(A))
+    @eval Base.:*(A::Adjoint{<:Any,<:$MAT}, D::IMatrix) = Adjoint(D * parent(A))
+    @eval Base.:*(A::Transpose{<:Any,<:$MAT}, D::IMatrix) = Transpose(D * parent(A))
     if MAT != :AbstactVector
-        @eval *(A::IMatrix, D::Transpose{<:Any,<:$MAT}) = Transpose(parent(D) * A)
-        @eval *(A::IMatrix, D::Adjoint{<:Any,<:$MAT}) = Adjoint(parent(D) * A)
+        @eval Base.:*(A::IMatrix, D::Transpose{<:Any,<:$MAT}) = Transpose(parent(D) * A)
```
Comment (Roger-luo) [inline src/linalg.jl:129]: perhaps worth checking if these can be done with `mul!` as well

**Review by Roger-luo — COMMENTED** [review]
we used to have some bad overloading in this package around IMatrix, not sure if all of them are fixed, better to double-check. I commented on some of them I spotted

File: src/linalg.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/linalg.jl:18]: Next time...

**Comment by GiggleLiu** [comment]
Inplace version is much more complicated. It requires checking the output matrix. The problem is we have too many matrix types now, we have `(# of matrix)^3` types of combination for the inplace interface instead of `(# of matrix)^2` in the non-inplace interface.

File: src/SSparseMatrixCSC.jl [hunk 3]
```diff
@@ -1,5 +1,3 @@
-export SSparseMatrixCSC
-
 @static if VERSION < v"1.4.0"
```
Comment (Roger-luo) [inline src/SSparseMatrixCSC.jl:1]: maybe just drop this? we don't really need to support 1.0 now

**Review by Roger-luo — APPROVED** [review]

**Pull request merged**